homeworks/solutions/krasnor/5-remesh/README.md

```diff
@@ -0,0 +1,70 @@
+# Homework 5: Quadric Error Metric Simplification
+
+In this homework, your task is to implement the
+[QEM Simplification](https://dl.acm.org/doi/abs/10.1145/258734.258849)
+algorithm.
+
+## Skeleton
+
+The skeleton for this homework is structured as follows:
+
+```
+5-remesh
+├── package.json
+├── package-lock.json
+├── README.md                <-- You work on this file
+├── src
+│   ├── assets
+│   │   ├── bunny_tri.obj
+│   │   └── sphere.obj
+│   ├── halfedge.js          <-- You work on this file
+│   ├── main.js              <-- Change this file if needed
+│   ├── mat.js
+│   ├── pq.js
+│   ├── renderer.js
+│   └── vec.js
+└── webpack.config.js
+```
+
+Implementation hints:
+
+1. Your primary task is to implement the `Halfedge.simplify` method in the `halfedge.js`, but you will have to build halfedge representation of a given mesh first. Reuse your work from previous homeworks
+2. Use `PriorityQueue` implementation from `pq.js` and `Matrix` implementation from `mat.js`
+3. It might be useful to read the code in `main.js`, and you are allowed to change code there if needed
+
+Think and answer the following questions:
+
+1. What's the most unanticipated issue when you were implementing QEM? Explain your coding experience and encountered challenges briefly.
+2. What are the cases missing from algorithm description, and how did you deal with it? For example, in which case you cannot do edge collapse directly. Explain your decision in your implementation.
+
+```
+1. Debugging is very difficult. Especially checking the results as no error values were known. I had to reread the paper multiple times and tried to follow the described algorithm as close as possible. In the end it looks like that I found something that seems to produce good error values for the bunny.
```

Comment:
I agree that debugging is difficult (including the matrix add method was wrong, this can really waste a lot of time). I try to implement it today spend a lot of time and finally traced it down to the error.